### ISSUE TITLE
Expected Memory reduction when using PEFT for flant5-XL

### ISSUE THREAD

**Issue by abdulvirta** [huggingface/peft#148]
Hi all, I'm trying to train google/flan-t5 xlarge on A100 40GB machines with pytorch lightning and deepspeed stage 2. Using AdamW as the optimizer. I'm using Lora config (r=8, alpha=32, dropout=0.1) and I can see the following:
trainable params: 4718592 || all params: 2854475776 || trainable%: 0.16530502867367827

However, my memory consumption on A100 40 GB is still around 34 GB with batch size 1. I'm not able to even fit batch_size=2. I expected a much more reduction in memory footprint with that percentage of trainable parameters. 

In terms of speed, I'm seeing around 5x increase in speed. Again expected much more significant increase in speed. Anything I'm missing to reduce the memory requirements to the level where this can be trained on standard colab machines as mentioned here:
https://huggingface.co/blog/peft

Any hints would be greatly appreciated!

**Comment by vladsavelyev** [comment]
Same question here. I'm training a modification GPT2-M on a google Colab GPU, using the `transformers.Trainer` class. With a standard fine-tuning (125,379,072 params), training takes ~20G of GPU RAM.

Now, with a peft Lora wrapper:

```LoraConfig(
        task_type=TaskType.CAUSAL_LM, 
        inference_mode=False, 
        r=4,
        lora_alpha=32,
        lora_dropout=0.1,
    )
```

It reports reducing the number of trainable params by a factor thousand:

```
trainable params: 147456 || all params: 125379072 || trainable%: 0.1176081443639972
```

However, when it comes to training, the GPU RAM usage is still exactly the same ~20G.

I manually verified that for all non-lora parameters `require_grad` is set to `False`, so this behavior puzzles me.

**Comment by vladsavelyev** [comment]
Still unaddressed.

**Comment by jxmorris12** [comment]
I've seen the same issue on my end as well. I noticed a lot of the documentation & tutorials use int8 quantization in addition to LORA. Not sure what speedup to expect with and without int8. But without int8 I see a slowdown and no memory decrease as well.

**Comment by pacman100** [comment]
Hello, it might be the case that the sequence lengths you use are longer, leading to bottlenecks in activations occupying high GPU memory. In that case, enable activation checkpointing and use DeepSpeed ZeRO-3 for sharding parameters/gradients/optimizer states across GPUs and if required DeepSpeed's CPU offloading. 

Not much we can do without reproducible script, the example scripts in example folder show the GPu memory occupied with DeepSpeed + CPU offloading for the raft datasets here: https://github.com/huggingface/peft/blob/main/examples/conditional_generation/peft_lora_seq2seq_accelerate_ds_zero3_offload.py

**Comment by pacman100** [comment]
Below are my rule of thumb:
1. Want to fit whole large LLM onto single GPU and reduce the model memory, use INT8 quantization and the preprocessing `prepare_for_int8_training()` also activates gradient checkpointing. So, model size as well as activations occupying GPU memory is reduced. With PEFT you have small number of trainable params now, so small optimizer states and less greadient computation limited to these. 

2. Have multi GPUs and don't want to use INT8 quantization. Use PEFT + DeepSpeed + Gradient Checkpointing. 

To quote the DeepSpeedChat repo: https://github.com/microsoft/DeepSpeedExamples/blob/master/applications/DeepSpeed-Chat/training/step1_supervised_finetuning/README.md
```
One important consideration for users is determining the maximum model size they can train using their current system. Here, we present a method for estimating this limit. Assuming that you do not use the offload feature and enable (i) zero stage 3 (if using multiple GPUs), (ii) gradient checkpoint, and (iii) LoRA, the approximate maximum model size (in billions of parameters) that you can train can be estimated as "Total GPU memory in GB divided by 3." For example, if you have a single A6000-48G GPU, you can probably train models up to 16 billion parameters. It is important to note that this is a rough estimation, and you should verify it by yourselves.
```

**Comment by jxmorris12** [comment]
this is still an issue!
